### ISSUE TITLE
zlibBufferSync: TypeError: Not a string or buffer

### ISSUE THREAD

**Issue by oldwizard1010** [oven-sh/bun#5309]
### What version of Bun is running?

1.0

### What platform is your computer?

Linux 6.2.16-2-arch1-1 x86_64 unknown

### What steps can reproduce the bug?

git clone https://github.com/ethereum-optimism/optimism-tutorial
cd ethereum-optimism/optimism-tutorial
npm install || bun install
mv ./.env.example ./.env

add, for example, https://1rpc.io/op to OPTIMISM_MAINNET_URL= in .env
then run:

node gas.js --network mainnet
bun gas.js --network mainnet

### What is the expected behavior?

```
node gas.js --network mainnet    INT ✘  10s  
<ref *1> Error: cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ] (error={"reason":"cannot estimate gas; transaction may fail or may require manual gas limit","code":"UNPREDICTABLE_GAS_LIMIT","error":{"reason":"processing response error","code":"SERVER_ERROR","body":"{\"jsonrpc\":\"2.0\",\"id\":48,\"error\":{\"code\":-32000,\"message\":\"gas required exceeds allowance (3475)\"}}\n","error":{"code":-32000},"requestBody":"{\"method\":\"eth_estimateGas\",\"params\":[{\"type\":\"0x2\",\"maxFeePerGas\":\"0x59682f68\",\"maxPriorityFeePerGas\":\"0x59682f00\",\"from\":\"0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266\",\"to\":\"0xcf210488dad6da5fe54d260c45253afc3a9e708c\",\"data\":\"0xa41368620000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000648656c6c6f210000000000000000000000000000000000000000000000000000\"}],\"id\":48,\"jsonrpc\":\"2.0\"}","requestMethod":"POST","url":"https://optimism.publicnode.com"},"method":"estimateGas","transaction":{"from":"0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266","maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x59682f00"},"maxFeePerGas":{"type":"BigNumber","hex":"0x59682f68"},"to":"0xCf210488DAd6da5FE54D260c45253AFC3A9E708C","data":"0xa41368620000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000648656c6c6f210000000000000000000000000000000000000000000000000000","type":2,"accessList":null}}, tx={"data":"0xa41368620000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000648656c6c6f210000000000000000000000000000000000000000000000000000","to":{},"from":"0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266","type":2,"maxFeePerGas":{"type":"BigNumber","hex":"0x59682f68"},"maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x59682f00"},"nonce":{},"gasLimit":{},"chainId":{}}, code=UNPREDICTABLE_GAS_LIMIT, version=abstract-signer/5.7.0)
    at Logger.makeError (node_modules/@ethersproject/logger/lib/index.js:238:21)
    at Logger.throwError (node_modules/@ethersproject/logger/lib/index.js:247:20)
    at node_modules/@ethersproject/abstract-signer/lib/index.js:365:47
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Promise.all (index 7) {
  reason: 'cannot estimate gas; transaction may fail or may require manual gas limit',
  code: 'UNPREDICTABLE_GAS_LIMIT',
  error: Error: cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ] (error={"reason":"processing response error","code":"SERVER_ERROR","body":"{\"jsonrpc\":\"2.0\",\"id\":48,\"error\":{\"code\":-32000,\"message\":\"gas required exceeds allowance (3475)\"}}\n","error":{"code":-32000},"requestBody":"{\"method\":\"eth_estimateGas\",\"params\":[{\"type\":\"0x2\",\"maxFeePerGas\":\"0x59682f68\",\"maxPriorityFeePerGas\":\"0x59682f00\",\"from\":\"0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266\",\"to\":\"0xcf210488dad6da5fe54d260c45253afc3a9e708c\",\"data\":\"0xa41368620000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000648656c6c6f210000000000000000000000000000000000000000000000000000\"}],\"id\":48,\"jsonrpc\":\"2.0\"}","requestMethod":"POST","url":"https://optimism.publicnode.com"}, method="estimateGas", transaction={"from":"0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266","maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x59682f00"},"maxFeePerGas":{"type":"BigNumber","hex":"0x59682f68"},"to":"0xCf210488DAd6da5FE54D260c45253AFC3A9E708C","data":"0xa41368620000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000648656c6c6f210000000000000000000000000000000000000000000000000000","type":2,"accessList":null}, code=UNPREDICTABLE_GAS_LIMIT, version=providers/5.7.2)
      at Logger.makeError (node_modules/@ethersproject/logger/lib/index.js:238:21)
      at Logger.throwError (node_modules/@ethersproject/logger/lib/index.js:247:20)
      at checkError (node_modules/@ethersproject/providers/lib/json-rpc-provider.js:175:16)
      at JsonRpcProvider.<anonymous> (node_modules/@ethersproject/providers/lib/json-rpc-provider.js:751:47)
      at step (node_modules/@ethersproject/providers/lib/json-rpc-provider.js:48:23)
      at Object.throw (node_modules/@ethersproject/providers/lib/json-rpc-provider.js:29:53)
      at rejected (node_modules/@ethersproject/providers/lib/json-rpc-provider.js:21:65)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
    reason: 'cannot estimate gas; transaction may fail or may require manual gas limit',
    code: 'UNPREDICTABLE_GAS_LIMIT',
    error: Error: processing response error (body="{\"jsonrpc\":\"2.0\",\"id\":48,\"error\":{\"code\":-32000,\"message\":\"gas required exceeds allowance (3475)\"}}\n", error={"code":-32000}, requestBody="{\"method\":\"eth_estimateGas\",\"params\":[{\"type\":\"0x2\",\"maxFeePerGas\":\"0x59682f68\",\"maxPriorityFeePerGas\":\"0x59682f00\",\"from\":\"0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266\",\"to\":\"0xcf210488dad6da5fe54d260c45253afc3a9e708c\",\"data\":\"0xa41368620000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000648656c6c6f210000000000000000000000000000000000000000000000000000\"}],\"id\":48,\"jsonrpc\":\"2.0\"}", requestMethod="POST", url="https://optimism.publicnode.com", code=SERVER_ERROR, version=web/5.7.1)
        at Logger.makeError (node_modules/@ethersproject/logger/lib/index.js:238:21)
        at Logger.throwError (node_modules/@ethersproject/logger/lib/index.js:247:20)
        at node_modules/@ethersproject/web/lib/index.js:313:32
        at step (node_modules/@ethersproject/web/lib/index.js:33:23)
        at Object.next (node_modules/@ethersproject/web/lib/index.js:14:53)
        at fulfilled (node_modules/@ethersproject/web/lib/index.js:5:58)
        at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
      reason: 'processing response error',
      code: 'SERVER_ERROR',
      body: '{"jsonrpc":"2.0","id":48,"error":{"code":-32000,"message":"gas required exceeds allowance (3475)"}}\n',
      error: [Error],
      requestBody: '{"method":"eth_estimateGas","params":[{"type":"0x2","maxFeePerGas":"0x59682f68","maxPriorityFeePerGas":"0x59682f00","from":"0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266","to":"0xcf210488dad6da5fe54d260c45253afc3a9e708c","data":"0xa41368620000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000648656c6c6f210000000000000000000000000000000000000000000000000000"}],"id":48,"jsonrpc":"2.0"}',
      requestMethod: 'POST',
      url: 'https://optimism.publicnode.com'
    },
    method: 'estimateGas',
    transaction: {
      from: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266',
      maxPriorityFeePerGas: [BigNumber],
      maxFeePerGas: [BigNumber],
      to: '0xCf210488DAd6da5FE54D260c45253AFC3A9E708C',
      data: '0xa41368620000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000648656c6c6f210000000000000000000000000000000000000000000000000000',
      type: 2,
      accessList: null
    }
  },
  tx: {
    data: '0xa41368620000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000648656c6c6f210000000000000000000000000000000000000000000000000000',
    to: Promise { '0xCf210488DAd6da5FE54D260c45253AFC3A9E708C' },
    from: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266',
    type: 2,
    maxFeePerGas: BigNumber { _hex: '0x59682f68', _isBigNumber: true },
    maxPriorityFeePerGas: BigNumber { _hex: '0x59682f00', _isBigNumber: true },
    nonce: Promise { 48 },
    gasLimit: Promise { <rejected> [Circular *1] },
    chainId: Promise { 10 }
  }
}
```

### What do you see instead?

```
[0.03ms] ".env"
2274 |           buf = Buffer2.concat(buffers, nread);
2275 |         buffers = [], engine.close(), callback(err, buf);
2276 |       }
2277 |     }
2278 | 
2279 |     function zlibBufferSync(engine, buffer) {
                                                         ^
TypeError: Not a string or buffer
      at zlibBufferSync (node:zlib:2279:54)
      at node_modules/@ethersproject/web/lib/geturl.js:78:48
      at endReadableNT (node:stream:2387:52)
      at processTicksAndRejections (:1:2602)
```

**Comment by enghamzasalem** [comment]
Same here ++

**Comment by tkroll** [comment]
Getting the same. The offending type is a `Uint8Array`.

**Comment by dr-noid** [comment]
Same here, this seems like the only thing preventing me from switching a project over to Bun completely.

**Comment by enghamzasalem** [comment]
bun is awesome for bugs ...just kidding

**Comment by mrobokit** [comment]
@oldwizard1010 Given that response.body is an instance of Uint8Array and still getting a TypeError, it's likely that zlib_1.gunzipSync expects a Node.js Buffer rather than a Uint8Array. Although these types are similar, they are not the same and some Node.js native modules expect Buffer objects.

getUrl.js, where you have the res.on

```
resp.on("end", function () {
    if (response.headers["content-encoding"] === "gzip") {
        //const size = response.body.length;
        response.body = (0, bytes_1.arrayify)((0, zlib_1.gunzipSync)(response.body));
        //console.log("Delta:", response.body.length - size, Buffer.from(response.body).toString());
    }
    resolve(response);
});
```
This is what i did to bypass the error:

```
const buffer = Buffer.from(response.body);
response.body = bytes_1.arrayify(zlib_1.gunzipSync(buffer));
```

Let me know if it worked in your case too!

**Comment by dr-noid** [comment]
@manzuni we probably have two different versions of zlib since my resp.on looks like:

```ts
resp.on("end", () => {
    if (headers["content-encoding"] === "gzip" && body) {
        body = getBytes(gunzipSync(body));
    }
    resolve({ statusCode, statusMessage, headers, body });
});
```

I changed it to `body = getBytes(gunzipSync(Buffer.from(body))` for now and so far I've not ran into that issue again.

**Comment by Ansonhkg** [comment]
If you can, ethers 6 seems not to have this problem

EDIT: Sorry I tried `6.7.1`

**Comment by enghamzasalem** [comment]
@Ansonhkg ethers 6 not workin

**Comment by tkroll** [comment]
> @manzuni we probably have two different versions of zlib since my resp.on looks like:
> 
> ```ts
> resp.on("end", () => {
>     if (headers["content-encoding"] === "gzip" && body) {
>         body = getBytes(gunzipSync(body));
>     }
>     resolve({ statusCode, statusMessage, headers, body });
> });
> ```
> 
> I changed it to `body = getBytes(gunzipSync(Buffer.from(body))` for now and so far I've not ran into that issue again.

This is working for me too.
Changed in node_modules/ethers/src.ts/utils/geturl.ts line 80. 
I'm using ethers 6.7.1.

**Comment by tkroll** [comment]
> @manzuni we probably have two different versions of zlib since my resp.on looks like:
> 
> ```ts
> resp.on("end", () => {
>     if (headers["content-encoding"] === "gzip" && body) {
>         body = getBytes(gunzipSync(body));
>     }
>     resolve({ statusCode, statusMessage, headers, body });
> });
> ```
> 
> I changed it to `body = getBytes(gunzipSync(Buffer.from(body))` for now and so far I've not ran into that issue again.

This is working for me too.
Changed in node_modules/ethers/lib.esm/utils/geturl.js line 65.
I'm using ethers 6.7.1.

Looking into the fn sig for `gunzipSync` it has `buffer` as `InputType` which includes `NodeJS.ArrayBufferView` which includes `TypedArray` which includes `Uint8Array`, so the Unit8Array is valid.

Looking into Bun's source: 
```
function zlibBufferSync(engine, buffer) {
    if (typeof buffer === "string") buffer = Buffer2.from(buffer);
    if (!Buffer2.isBuffer(buffer)) throw new TypeError("Not a string or buffer");
    var flushFlag = engine._finishFlushFlag;
    return engine._processChunk(buffer, flushFlag);
}
```

Looks like it needs to call `Buffer2.from(buffer)` for anything not a Buffer.

**Comment by Jarred-Sumner** [comment]
Our node:zlib polyfill is not implemented well. We need to make it use the native implementation in Bun. It currently uses a JS implementation intended for browsers.

**Comment by kartojal** [comment]
in the meantime published a hot fix to bypass bun bug in `ethers` package, and you can install as alias with bun, if you feel adventurous:

```
bun install ethers@npm:fix-geturl-ethers
```

⚠️ trusting random NPM packages is bad, please be cautious, changes here ⚠️

https://github.com/ethers-io/ethers.js/compare/main...kartojal:ethers.js:main#diff-40e9e0d3db93684f29c405aeee3b29cbbb19419dc79bf1c9e602199c93abf7c1R80

**Comment by K-Mistele** [comment]
^^ that closed issue was mine, same issue here

**Comment by dmytrotkk** [comment]
> @manzuni we probably have two different versions of zlib since my resp.on looks like:
> 
> ```ts
> resp.on("end", () => {
>     if (headers["content-encoding"] === "gzip" && body) {
>         body = getBytes(gunzipSync(body));
>     }
>     resolve({ statusCode, statusMessage, headers, body });
> });
> ```
> 
> I changed it to `body = getBytes(gunzipSync(Buffer.from(body))` for now and so far I've not ran into that issue again.

That helped, thanks!
But I can't find an open ticket in ethers repo for that. Is it going to be fixed on ethers side or in bun?

**Comment by harrysupertramp** [comment]
If you want an already fixed ethers.js (v6.9.0) version try this fork: 
`bun add ShadowyCreators/ethers.js`

https://github.com/ShadowyCreators/ethers.js

**Comment by cah4a** [comment]
@dmytrotkk, The node version of `gunzipSync` works well with UInt8Array. 
Ethers.js is not the one relying on this, so, in my opinion, it is better to fix it on bun side.

**Comment by cah4a** [comment]
One more way to work around this is sed the bug out.
Put this postinstall script into your package json:
```json
{
    "scripts": {
        "postinstall": "find node_modules/ethers -type f -name 'geturl.js' -exec sed -i '' 's/getBytes(gunzipSync(body))/getBytes(gunzipSync(Buffer.from(body)))/g' {} +"
    }
}
```

**Comment by reverie** [comment]
> One more way to work around this is sed the bug out. Put this postinstall script into your package json:
> 
> ```json
> {
>     "scripts": {
>         "postinstall": "find node_modules/ethers -type f -name 'geturl.js' -exec sed -i '' 's/getBytes(gunzipSync(body))/getBytes(gunzipSync(Buffer.from(body)))/g' {} +"
>     }
> }
> ```

Thanks! This worked for me, I just had to mod it for the commonjs version as well

**Comment by cah4a** [comment]
@reverie For Dockerfile support I've made following .sh file instead of inline command:
```sh
if [[ "$OSTYPE" == "darwin"* ]]; then
  find node_modules/ethers -type f -name 'geturl.js' -exec sed -i '' 's/getBytes(gunzipSync(body))/getBytes(gunzipSync(Buffer.from(body)))/g' {} +
else
  find node_modules/ethers -type f -name 'geturl.js' -exec sed -i 's/getBytes(gunzipSync(body))/getBytes(gunzipSync(Buffer.from(body)))/g' {} +
fi
```

**Comment by Electroid** [comment]
This should be fixed, it will be available in Bun v1.0.23. You can try it now by running:
`bun upgrade --canary`
